staging/src/k8s.io/kubectl/pkg/cmd/debug/profiles.go

```diff
@@ -49,11 +62,195 @@ func profileLegacy(pod *corev1.Pod, containerName string, target runtime.Object)
+			c.VolumeMounts = append(c.VolumeMounts, corev1.VolumeMount{
+				MountPath: "/host",
+				Name:      volumeName,
+			})
+		})
+		pod.Spec.SecurityContext = nil
+		setHostNamespace(pod, true)
+	default:
+		return fmt.Errorf("the %s profile doesn't support objects of type %T", ProfileGeneral, target)
+	}
+
+	return nil
+}
+
+func (p *baselineProfile) Apply(pod *corev1.Pod, containerName string, target runtime.Object) error {
+	switch getDebugStyle(pod, target) {
+	case stylePodCopy:
+		// For copy of pod: empty securityContext; sets shareProcessNamespace
+		pod.Spec.SecurityContext = nil
+		pod.Spec.ShareProcessNamespace = pointer.BoolPtr(true)
+		setHostNamespace(pod, false)
+		modifyContainer(pod.Spec.Containers, containerName, func(c *corev1.Container) {
+			c.SecurityContext = nil
+			c.SecurityContext = addCap(c.SecurityContext, "SYS_PTRACE")
+		})
+
+	case styleEphemeral:
+		for i := range pod.Spec.EphemeralContainers {
+			container := &pod.Spec.EphemeralContainers[i]
+			if container.Name != containerName {
+				continue
+			}
+			// For ephemeral container: empty securityContext
+			container.SecurityContext = nil
+		}
+
+	case styleNode:
+		// empty securityContext; uses isolated namespaces
+		pod.Spec.SecurityContext = nil
+		setHostNamespace(pod, false)
+
+	default:
+		return fmt.Errorf("the %s profile doesn't support objects of type %T", ProfileBaseline, target)
+	}
+
+	return nil
+}
+
+func (p *restrictedProfile) Apply(pod *corev1.Pod, containerName string, target runtime.Object) error {
+	// drops privilege and run as non-root
+	sc := &corev1.SecurityContext{
+		RunAsNonRoot: pointer.BoolPtr(true),
+		Capabilities: &corev1.Capabilities{
+			Drop: []corev1.Capability{"ALL"},
+		},
+	}
+
+	switch getDebugStyle(pod, target) {
+	case stylePodCopy:
+		// For copy of pod: empty securityContext; sets shareProcessNamespace
+		pod.Spec.ShareProcessNamespace = pointer.BoolPtr(true)
+		modifyContainer(pod.Spec.Containers, containerName, func(c *corev1.Container) {
+			c.SecurityContext = sc
+		})
+
+	case styleEphemeral:
+		modifyEphemeralContainer(pod.Spec.EphemeralContainers, containerName, func(c *corev1.EphemeralContainer) {
+			c.SecurityContext = sc
+		})
+
+	case styleNode:
+		// no additional settings required other than the common
+		// settings applied below
+
+	default:
+		return fmt.Errorf("the %s profile doesn't support objects of type %T", ProfileRestricted, target)
+	}
+
+	// common settings, empty securityContext and uses isolated namespaces
+	pod.Spec.SecurityContext = nil
+	setHostNamespace(pod, false)
+
+	return nil
+}
+
+func setHostNamespace(pod *corev1.Pod, enabled bool) {
+	pod.Spec.HostNetwork = enabled
+	pod.Spec.HostPID = enabled
+	pod.Spec.HostIPC = enabled
+}
+
+func addCap(s *corev1.SecurityContext, c corev1.Capability) *corev1.SecurityContext {
+	if s == nil {
+		return &corev1.SecurityContext{
+			Capabilities: &corev1.Capabilities{
+				Add: []corev1.Capability{c},
+			},
+		}
+	}
+
+	ss := s.DeepCopy()
```

Comment:
Because most profiles will need to do this, it helps with readability and consistency to provide a helper function. It's very easy to mess this up and induce a panic if we expect each profile to implement this due to `nil` pointers.

We don't technically need to do the deep copy, I was only doing to have it so that the input argument `s` is not mutated. I've removed the deep copy.

To further improve the readability at call sites, I've renamed this from `addCap` to `addCapability`.

This is all done in 1b9e3ad05be421f0d3e0949e03f2d198a3f128ac